In [53]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [54]:
df = pd.read_csv('student_data.csv')

In [55]:
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

In [57]:
df.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  395.000000  395.000000  395.000000  395.000000  
mean     5.708861   10.908861   10.713924   10.415190  
std      8.003096    3.319195    3.761505    4.581443  
min      0.000000    3.000000    0.000000    0.000000  
25%      0.000000    8.000000    9.000000    8.000000  
50%      4.000000   11.000000   11.000000   11.000000  
75%      8.000000   13.000000   13.000000   14.000000  
max     75.000000   19.000000   19.000000   20.000000

In [58]:
df.isnull().value_counts()


school  sex    age    address  famsize  Pstatus  Medu   Fedu   Mjob   Fjob   reason  guardian  traveltime  studytime  failures  schoolsup  famsup  paid   activities  nursery  higher  internet  romantic  famrel  freetime  goout  Dalc   Walc   health  absences  G1     G2     G3   
False   False  False  False    False    False    False  False  False  False  False   False     False       False      False     False      False   False  False       False    False   False     False     False   False     False  False  False  False   False     False  False  False    395
Name: count, dtype: int64

In [76]:
# no null values in the data set
# all rows are unique
df['student_id'] = range(1,len(df)+1)

In [77]:
df.duplicated().sum()

np.int64(0)

In [78]:
# Move 'student_id' to the first column
df = df[['student_id'] + [col for col in df.columns if col != 'student_id']]


In [79]:
df

student_id school sex  age address famsize Pstatus  Medu  Fedu      Mjob  \
0             1     GP   F   18       U     GT3       A     4     4   at_home   
1             2     GP   F   17       U     GT3       T     1     1   at_home   
2             3     GP   F   15       U     LE3       T     1     1   at_home   
3             4     GP   F   15       U     GT3       T     4     2    health   
4             5     GP   F   16       U     GT3       T     3     3     other   
..          ...    ...  ..  ...     ...     ...     ...   ...   ...       ...   
390         391     MS   M   20       U     LE3       A     2     2  services   
391         392     MS   M   17       U     LE3       T     3     1  services   
392         393     MS   M   21       R     GT3       T     1     1     other   
393         394     MS   M   18       R     LE3       T     3     2  services   
394         395     MS   M   19       U     LE3       T     1     1     other   

     ... famrel freetime goout  Dalc  Walc  health absences  G1  G2  G3  
0    ...      4        3     4     1     1       3        6   5   6   6  
1    ...      5        3     3     1     1       3        4   5   5   6  
2    ...      4        3     2     2     3       3       10   7   8  10  
3    ...      3        2     2     1     1       5        2  15  14  15  
4    ...      4        3     2     1     2       5        4   6  10  10  
..   ...    ...      ...   ...   ...   ...     ...      ...  ..  ..  ..  
390  ...      5        5     4     4     5       4       11   9   9   9  
391  ...      2        4     5     3     4       2        3  14  16  16  
392  ...      5        5     3     3     3       3        3  10   8   7  
393  ...      4        4     1     3     4       5        0  11  12  10  
394  ...      3        2     3     3     3       5        5   8   9   9  

[395 rows x 34 columns]

Summary of the dataset

this data set is of size 395 * 33

it consist of various information regarding a a student such as its personal details,academic qualifications etc

by analyzing all the information we have to figure out that which all factors affect the performance of students academically
and what better the student could do so as to excel in his/her career

all rows are unique


In [80]:
df

student_id school sex  age address famsize Pstatus  Medu  Fedu      Mjob  \
0             1     GP   F   18       U     GT3       A     4     4   at_home   
1             2     GP   F   17       U     GT3       T     1     1   at_home   
2             3     GP   F   15       U     LE3       T     1     1   at_home   
3             4     GP   F   15       U     GT3       T     4     2    health   
4             5     GP   F   16       U     GT3       T     3     3     other   
..          ...    ...  ..  ...     ...     ...     ...   ...   ...       ...   
390         391     MS   M   20       U     LE3       A     2     2  services   
391         392     MS   M   17       U     LE3       T     3     1  services   
392         393     MS   M   21       R     GT3       T     1     1     other   
393         394     MS   M   18       R     LE3       T     3     2  services   
394         395     MS   M   19       U     LE3       T     1     1     other   

     ... famrel freetime goout  Dalc  Walc  health absences  G1  G2  G3  
0    ...      4        3     4     1     1       3        6   5   6   6  
1    ...      5        3     3     1     1       3        4   5   5   6  
2    ...      4        3     2     2     3       3       10   7   8  10  
3    ...      3        2     2     1     1       5        2  15  14  15  
4    ...      4        3     2     1     2       5        4   6  10  10  
..   ...    ...      ...   ...   ...   ...     ...      ...  ..  ..  ..  
390  ...      5        5     4     4     5       4       11   9   9   9  
391  ...      2        4     5     3     4       2        3  14  16  16  
392  ...      5        5     3     3     3       3        3  10   8   7  
393  ...      4        4     1     3     4       5        0  11  12  10  
394  ...      3        2     3     3     3       5        5   8   9   9  

[395 rows x 34 columns]

In [81]:
df.drop('student_id',axis=1,inplace=True)

C:\Users\dell\AppData\Local\Temp\ipykernel_12240\542718699.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop('student_id',axis=1,inplace=True)


In [75]:
df

school sex  age address famsize Pstatus  Medu  Fedu      Mjob      Fjob  \
0       GP   F   18       U     GT3       A     4     4   at_home   teacher   
1       GP   F   17       U     GT3       T     1     1   at_home     other   
2       GP   F   15       U     LE3       T     1     1   at_home     other   
3       GP   F   15       U     GT3       T     4     2    health  services   
4       GP   F   16       U     GT3       T     3     3     other     other   
..     ...  ..  ...     ...     ...     ...   ...   ...       ...       ...   
390     MS   M   20       U     LE3       A     2     2  services  services   
391     MS   M   17       U     LE3       T     3     1  services  services   
392     MS   M   21       R     GT3       T     1     1     other     other   
393     MS   M   18       R     LE3       T     3     2  services     other   
394     MS   M   19       U     LE3       T     1     1     other   at_home   

     ... famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0    ...      4        3      4     1     1      3        6   5   6   6  
1    ...      5        3      3     1     1      3        4   5   5   6  
2    ...      4        3      2     2     3      3       10   7   8  10  
3    ...      3        2      2     1     1      5        2  15  14  15  
4    ...      4        3      2     1     2      5        4   6  10  10  
..   ...    ...      ...    ...   ...   ...    ...      ...  ..  ..  ..  
390  ...      5        5      4     4     5      4       11   9   9   9  
391  ...      2        4      5     3     4      2        3  14  16  16  
392  ...      5        5      3     3     3      3        3  10   8   7  
393  ...      4        4      1     3     4      5        0  11  12  10  
394  ...      3        2      3     3     3      5        5   8   9   9  

[395 rows x 33 columns]

In [82]:
df[df.duplicated()]

Empty DataFrame
Columns: [school, sex, age, address, famsize, Pstatus, Medu, Fedu, Mjob, Fjob, reason, guardian, traveltime, studytime, failures, schoolsup, famsup, paid, activities, nursery, higher, internet, romantic, famrel, freetime, goout, Dalc, Walc, health, absences, G1, G2, G3]
Index: []

[0 rows x 33 columns]

Data quality Dimensions

completness -> is data missing
Validity -> Is dat invalid -> negative height -> duplicate patient id.
Accuracy -> data is valid but not accurate.
Consistency -> both valid and accurate but written differently -> New York and NY

Order of severity

completness <- Validity <- Accuracy <- Consistency

In [83]:
df['school'].value_counts()

school
GP    349
MS     46
Name: count, dtype: int64

In [85]:
df['sex'].value_counts()

sex
F    208
M    187
Name: count, dtype: int64

In [94]:
df[df['age'].isnull()]

Empty DataFrame
Columns: [school, sex, age, address, famsize, Pstatus, Medu, Fedu, Mjob, Fjob, reason, guardian, traveltime, studytime, failures, schoolsup, famsup, paid, activities, nursery, higher, internet, romantic, famrel, freetime, goout, Dalc, Walc, health, absences, G1, G2, G3]
Index: []

[0 rows x 33 columns]

In [95]:
df['address'].value_counts()

address
U    307
R     88
Name: count, dtype: int64

In [100]:
df['famsize'].value_counts()


famsize
GT3    281
LE3    114
Name: count, dtype: int64

In [101]:
df['Medu'].value_counts()

Medu
4    131
2    103
3     99
1     59
0      3
Name: count, dtype: int64

In [103]:
df['Fedu'].value_counts()

Fedu
2    115
3    100
4     96
1     82
0      2
Name: count, dtype: int64

In [106]:
df.isnull().sum()

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64

In [107]:
df['G1'].value_counts()

G1
10    51
8     41
11    39
7     37
12    35
13    33
9     31
14    30
15    24
6     24
16    22
18     8
17     8
5      7
19     3
4      1
3      1
Name: count, dtype: int64